Features - Add asymmetric links, add moving nodes, refactor for maintainability

medentem opened the pull request:
Added the ability to test the effect of utilizing  a bloom filter to manage routing rather than roles and delays currently used by the managed flood approach.

GUVWAF commented on lib/config.py:
```diff
@@ -46,7 +47,7 @@
 CADMODEM = np.array([-124.5, -127.0, -129.5, -132.0, -134.5, -137.5, -140.0, -143.0])
 FREQ = REGION["freq_start"]+BWMODEM[MODEM]*CHANNEL_NUM
-HEADERLENGTH = 16  # number of Meshtastic header bytes 
+HEADERLENGTH = 32  # number of Meshtastic header bytes 
```
This should be dependent on the `ROUTER_TYPE`.

medentem commented:
Added asymmetric link modeling by varying loss on either end of the link.

medentem commented on loraMesh.py:
```diff
@@ -145,6 +210,12 @@ def receive(self, in_pipe):
+				if not self.hasReceivedAnyPacket:
+					self.hasReceivedAnyPacket = True
+
+				# Update knowledge of node based on reception of packet
```
Resolved.

GUVWAF replied:
`self.updateCoverageKnowledge(p.txNodeId)` is now still always called, right?

medentem replied:
No.. it should only be called under the condition you described.

GUVWAF replied:
I commented on this when you only pushed the commit to not change `hasReceivedAnyPacket`, but it's irrelevant now that `relay_node` is extended.

medentem commented:
Position is better simulated now using the same exponential distribution as packet emission.

medentem commented:
Latest change is to move the MeshNode class out of loraMesh and batchSim... it was copy pasted in both place which makes it very easy to miss something between the two spots.

medentem commented:
Updating with better outputs. Single graph with line per router. Added % change callouts. Added % complete while simulation runs so its not like watching paint dry :-). Added time remaining to each simulation.

medentem commented:
<img width="704" alt="Screenshot 2025-01-13 at 6 02 58 PM" src="https://github.com/user-attachments/assets/05ecf18c-bd45-4685-99fe-1c1a533cc44b" />

medentem commented:
@GUVWAF  - one thing I could use your help with. For some reason, whichever ROUTER_TYPE is first in the routerTypes array in batchSim.py performs better and I am not understanding why. My only hypothesis is that as stats are accumulating in memory, the simulator runs more slowly which makes the second router type look like its worse. 

For the time being, I can just run one at a time, but I'm worried something is skewing the results here.

GUVWAF commented:
@medentem  Yeah, I remember having similar issues before, I think it has something to do with how `config.py` is imported. 

Since I couldn’t figure it out at that point, I created the separate `loraMesh.py` and `batchSim.py` files with duplicated code, which solved the issue for me.

medentem commented:
Thank you. I'll take a peek at that.

medentem commented:
I moved the config to an object with new instances per iteration. That seems to have solved it. But now I need to rerun all my tests :-D

medentem commented:
Squash to one commit. Removed bloom filter. Left the asymmetric links, moving nodes and some of the refactoring to increase maintainability (moved MeshNode to node.py instead of duplicating). Moved config to class to fix issues with batchSim affecting multiple run results. 

I left the structure intact for multiple ROUTER_TYPES for future testing. I may just take another crack at some new concept in the future :-D

GUVWAF commented:
Thank you for this, I will review and test in the coming days.

GUVWAF commented on lib/interactive.py:
```diff
@@ -4,18 +4,20 @@
 import sys
 import threading
 import time
+import base64
```
This import is unnecessary.

GUVWAF commented on lib/config.py:
```diff
@@ -1,66 +1,132 @@
+
+        # Misc
+        self.SEED = 44  # random seed to use
+        self.PLOT = False
```
Please keep `PLOT` set to `True` by default.

GUVWAF commented on batchSim.py:
```diff
@@ -1,357 +1,615 @@
 plt.ylabel('Collision rate (%)')
+plt.legend()
+plt.title('Collision Rate by Router Type (with % Diff Annotations)')
+plt.show()
```
I'd prefer to show all plots at once, so only one `plt.show()` at the end.

GUVWAF commented on lib/common.py:
```diff
@@ -1,3 +1,4 @@
+import math
```
Unnecessary import.

GUVWAF commented on lib/common.py:
```diff
@@ -6,8 +7,7 @@
 from matplotlib.widgets import Button, Slider, RadioButtons, TextBox
-
-from . import config as conf
+from lib.config import Config
```
Unnecessary import.

GUVWAF reviewed (commented):
This is really nice, not only the additions but also the general refactor, thanks a lot! 

Left a few minor comments. When those are addressed, I think it's good to merge.

medentem commented:
All set!

GUVWAF commented:
Merged!
